File: Week_03/powx_n/Solution.java
package powx_n;

/**
 * 暴力法
 * 时间复杂度: O(n)
 * 空间复杂度: O(1)
 */
class Solution1 {
    public double pow(double x, int n) {
        int res = 1;
        while (n-- > 0) {
            res *= x;
        }
        return res;
    }
}

/**
 * 分治思想
 * 时间复杂度: O(logn)，递归的层数
 * 空间复杂度: O(logn)，递归使用的栈空间
 */
class Solution2 {
    public double myPow(double x, int n) {
        long b = n; //int 的范围是 [-2^31, 2^31-1]，当 n = -2^31, 执行 n = -n 会溢出，所以转成 long
        return n > 0 ? pow(x, n) : 1 / pow(x, -b);
    }

    private double pow(double x, long b) {
        if (b == 0) return 1;
        double y = pow(x, b >> 1); //b >> 1 = b / 2
        return (b & 1) == 0 ? y * y : y * y * x; // b & 1 = b % 2
    }
}

File: Week_03/powx_n/Solution.py
# 递归
# 时间复杂度：O(logn)
# 空间复杂度：O(logn)
class Solution1:
    def myPow(self, x: float, n: int) -> float:
        # 传正数 n
        def pow(n):
            if n == 0: return 1
            y = pow(n >> 1)
            return y * y if n & 1 == 0 else y * y * x

        # 处理 n 如果是负数
        return pow(n) if n >= 0 else 1 / pow(-n)


# 循环
# 时间复杂度：O(logn)
# 空间复杂度：O(1)
class Solution2:
    def myPow(self, x: float, n: int) -> float:
        if x == 0.0: return 0.0
        res = 1
        if n < 0: x, n = 1 / x, -n
        while n:
            if n & 1: res *= x
            x *= x
            n >>= 1
        return res
